TestSingleKey: no txn persisted, yet intent has heartbeat

The following test appears to have failed:

[#5940](https://circleci.com/gh/cockroachdb/cockroach/5940):

```
=== RUN TestPut
--- PASS: TestPut (22.56s)
=== RUN TestRangeReplication
--- PASS: TestRangeReplication (12.07s)
=== RUN TestSingleKey
--- FAIL: TestSingleKey (19.64s)
	single_key_test.go:125: txn "acceptance/single_key_test.go:101 func·008" id=d508bfcc key="test-key" rw=false pri=87.92241271 iso=SERIALIZABLE stat=PENDING epo=1 ts=1439565620.755168676,2 orig=1439565620.742974281,0 max=1439565620.992974043,0: no txn persisted, yet intent has heartbeat
=== RUN TestStatusServer
--- PASS: TestStatusServer (18.48s)
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -743,6 +743,24 @@ func (r *Replica) PushTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.
 	if err != nil {
 		return reply, err
 	}
+	// There are three cases in which there is no transaction entry, in
+	// decreasing likelihood:
+	// * the pushee is still active; it just hasn't committed, restarted,
+	//   aborted or heartbeat yet.
+	// * the pushee resolved its intents synchronously on successful commit;
+	//   in this case, the transaction record of the pushee is also removed.
+	//   Note that in this case, the intent which prompted this PushTxn
+	//   doesn't exist any more.
+	// * the pushee timed out or was aborted and the intent not cleaned up,
+	//   but the transaction record garbage collected.
+	//
+	// We currently make no attempt at guessing which one it is, though we
+	// could (see #1939). Instead, a new entry is always written.
+	//
+	// TODO(tschottdorf): we should actually improve this when we
+	// garbage-collect aborted transactions, or we run the risk of a push
+	// recreating a GC'ed transaction as PENDING, which is an error if it
+	// has open intents (which is likely if someone pushes it).
 	if ok {
 		// Start with the persisted transaction record as final transaction.
 		reply.PusheeTxn = gogoproto.Clone(existTxn).(*proto.Transaction)
@@ -756,10 +774,7 @@ func (r *Replica) PushTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.
 		}
 	} else {
 		// Some sanity checks for case where we don't find a transaction record.
-		if args.PusheeTxn.LastHeartbeat != nil {
-			return reply, proto.NewTransactionStatusError(&args.PusheeTxn,
-				"no txn persisted, yet intent has heartbeat")
-		} else if args.PusheeTxn.Status != proto.PENDING {
+		if args.PusheeTxn.Status != proto.PENDING {
 			return reply, proto.NewTransactionStatusError(&args.PusheeTxn,
 				fmt.Sprintf("no txn persisted, yet intent has status %s", args.PusheeTxn.Status))
 		}

